### PR TITLE
Links Awakening: Implement New Game

### PR DESCRIPTION
**Pull request by zig-for**
## What is this fixing or adding?
Adds Link's Awakening: DX. Fully imports and forks LADXR, with permission - https://github.com/daid/LADXR

## How was this tested?
Multiple beta tests. No major issues have been revealed over the past few ones.

## Known bugs:
* Check count at the end screen could be off by 5, due to local only items using a special snowflake collection method - will fix in later revision
* Client hangs when waiting on player to actually collect an item
* One person was unable to connect to Retroarch, still a mystery why, need more data

### CODE REVIEW

File: worlds/ladx/__init__.py [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+        "English",
+        "setup_en.md",
+        "setup/en",
+        ["zig"]
+    )]
+    theme = "dirt"
+
+class LinksAwakeningWorld(World):
+    """Insert description of the world/game here."""
+    game: str = LINKS_AWAKENING # name of the game/world
```
Comment (Jarno458) [inline worlds/ladx/__init__.py:39]: i am not sure where you getting this value from `LINKS_AWAKENING`

File: worlds/ladx/LADXR/.github/workflows/multiworld_build.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
+name: Build multiworld client
```
Comment (Jarno458) [inline worlds/ladx/LADXR/.github/workflows/multiworld_build.yml:1]: I dont think you need to include this build pipeline

File: worlds/ladx/LADXR/.gitignore [hunk 3]
```diff
@@ -0,0 +1,143 @@
+# Byte-compiled / optimized / DLL files
```
Comment (Jarno458) [inline worlds/ladx/LADXR/.gitignore:1]: I dont think gitignore works in subfolders

File: worlds/ladx/LADXR/LICENSE [hunk 4]
```diff
@@ -0,0 +1,21 @@
+MIT License
```
Comment (Jarno458) [inline worlds/ladx/LADXR/LICENSE:1]: It should be clear this license is for LADXR or just fall under the MIT of the AP project

File: worlds/ladx/LADXR/.gitignore (hunk 3 above)
Comment (zig-for) [reply worlds/ladx/LADXR/.gitignore:1]: gitignore does work in subfolders! But this one is probably not needed

File: worlds/ladx/LADXR/LICENSE (hunk 4 above)
Comment (zig-for) [reply worlds/ladx/LADXR/LICENSE]: addressed

File: worlds/ladx/LADXR/.gitignore (hunk 3 above)
Comment (zig-for) [reply worlds/ladx/LADXR/.gitignore]: addressed

File: worlds/ladx/LADXR/.github/workflows/multiworld_build.yml (hunk 2 above)
Comment (zig-for) [reply worlds/ladx/LADXR/.github/workflows/multiworld_build.yml]: addressed

**Review by Jarno458 — COMMENTED** [review]
Tnx for the cleanup

File: worlds/ladx/docs/en_Links Awakening DX.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+contain any of those items may have their contents changed.
+
+## Which items can be in another player's world?
+
+Any of the items which can be shuffled may also be placed into another player's world. It is possible to choose to limit
+certain items to your own world.
+
+## What does another world's item look like in Link's Awakening?
+
+The game will try to pick an appropriate sprite for the item (a LttP sword will be a sword!) - it may however be a little odd (a Missile Pack may be a single arrow).
```
Comment (SoldierofOrder) [inline worlds/ladx/docs/en_Links Awakening DX.md:26]:
```suggestion
The game will try to pick an appropriate sprite for the item (a LttP sword will be a sword!) - it may, however, be a little odd (a Missile Pack may be a single arrow).
```

File: worlds/ladx/docs/setup_en.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
+# Links Awakening DX Multiworld Setup Guide
+
+## Required Software
+
+- [Archipelago](https://github.com/ArchipelagoMW/Archipelago/releases). Make sure to check the box for `Links Awakening DX`
+- Hardware or software capable of loading and playing SNES ROM files
+    - Currently only [RetroArch](https://retroarch.com?page=platforms) 1.10.3 or newer) is supported
+    - Bizhawk support will come at a later date
```
Comment (SoldierofOrder) [inline worlds/ladx/docs/setup_en.md:8]:
```suggestion
    - Currently only [RetroArch](https://retroarch.com?page=platforms) 1.10.3 or newer) is supported.
    - Bizhawk support will come at a later date.
```

File: worlds/ladx/docs/setup_en.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+[YAML Validator](/mysterycheck) page.
+
+## Generating a Single-Player Game
+
+1. Navigate to the [Player Settings](/games/Links%20Awakening%20DX/player-settings) page, configure your options,
+   and click the "Generate Game" button.
+2. You will be presented with a "Seed Info" page.
+3. Click the "Create New Room" link.
+4. You will be presented with a server page, from which you can download your patch file.
+5. Double-click on your patch file, and Links Awakening DX will launch automatically, and create your ROM from the patch file
```
Comment (SoldierofOrder) [inline worlds/ladx/docs/setup_en.md:52]:
```suggestion
5. Double-click on your patch file, and Links Awakening DX will launch automatically, and create your ROM from the patch file.
```

File: worlds/ladx/docs/setup_en.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+   Network Command Port at 55355.
+
+![Screenshot of Network Commands setting](/static/generated/docs/A%20Link%20to%20the%20Past/retroarch-network-commands-en.png)
+4. Go to Main Menu --> Online Updater --> Core Downloader. Scroll down and select "Nintendo - SNES / SFC (bsnes-mercury
+   Performance)".
+
+### Connect to the Archipelago Server
+
+The patch file which launched your client should have automatically connected you to the AP Server. There are a few
+reasons this may not happen however, including if the game is hosted on the website but was generated elsewhere. If the
```
Comment (SoldierofOrder) [inline worlds/ladx/docs/setup_en.md:84]:
```suggestion
reasons this may not happen, however, including if the game is hosted on the website but was generated elsewhere. If the
```

**Review by SoldierofOrder — APPROVED** [review]
Only examined the two user-facing docs files (setup guide and info page). Very thorough and clear writing!

File: worlds/ladx/Items.py [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+ladxr_item_to_la_item_name = {
+    item.ladxr_id: item.item_name for item in links_awakening_items
+}
+
+links_awakening_items_by_name = {
+    item.item_name : item for item in links_awakening_items
+}
+
+for item in links_awakening_items_by_name.values():
+    print(item.item_name, Common.BASE_ID + item.item_id)
```
Comment (Alchav) [inline worlds/ladx/Items.py:307]: Debug code left in?

Comment (zig-for) [reply worlds/ladx/Items.py:307]: dangit yes, good catch

Comment (zig-for) [reply worlds/ladx/Items.py]: fixed

File: worlds/ladx/Options.py [hunk 10]
```diff
@@ -319,20 +320,19 @@ class GfxMod(FreeText, LADXROption):
...
 
-    __spriteFiles: typing.Dict[str, typing.List[str]] = {}
+    __spriteFiles: typing.DefaultDict[str, typing.List[str]] = defaultdict(list)
     __spriteDir = os.path.join('data', 'sprites','ladx')
 
+    extensions = [".bin", ".bdiff", ".png", ".bmp"]
     def __init__(self, value: str):
         super().__init__(value)
         if not GfxMod.__spriteFiles:
+            os.path.splitext
```
Comment (Berserker66) [inline worlds/ladx/Options.py:330]: this line doesn't do anything?

Comment (zig-for) [reply worlds/ladx/Options.py:330]: damnit - yeah

Comment (zig-for) [reply worlds/ladx/Options.py]: fixed

File: host.yaml [hunk 11]
```diff
@@ -82,7 +82,7 @@ generator:
   race: 0
   # List of options that can be plando'd. Can be combined, for example "bosses, items"
   # Available options: bosses, items, texts, connections
-  plando_options: "bosses"
+  plando_options: "items"
```
Comment (Berserker66) [inline host.yaml:85]: why is this change here?

Comment (zig-for) [reply host.yaml:85]: Had to change it during a beta test, accidentally committed. Will do another pass tonight for small things like this I've missed.

File: worlds/ladx/docs/setup_en.md [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+
+You only have to do these steps once. Note, RetroArch 1.9.x will not work as it is older than 1.10.3.
+
+1. Enter the RetroArch main menu screen.
+2. Go to Settings --> User Interface. Set "Show Advanced Settings" to ON.
+3. Go to Settings --> Network. Set "Network Commands" to ON. (It is found below Request Device 16.) Leave the default
+   Network Command Port at 55355.
+
+![Screenshot of Network Commands setting](/static/generated/docs/A%20Link%20to%20the%20Past/retroarch-network-commands-en.png)
+4. Go to Main Menu --> Online Updater --> Core Downloader. Scroll down and select "Nintendo - SNES / SFC (bsnes-mercury
```
Comment (Berserker66) [inline worlds/ladx/docs/setup_en.md:78]: This seems like the wrong core

File: inno_setup.iss [hunk 13]
```diff
@@ -219,6 +221,11 @@ Root: HKCR; Subkey: "{#MyAppName}pkmnbpatch";                     ValueData: "Ar
 Root: HKCR; Subkey: "{#MyAppName}pkmnbpatch\DefaultIcon";         ValueData: "{app}\ArchipelagoPokemonClient.exe,0";                           ValueType: string;  ValueName: ""; Components: client/pkmn
 Root: HKCR; Subkey: "{#MyAppName}pkmnbpatch\shell\open\command";  ValueData: """{app}\ArchipelagoPokemonClient.exe"" ""%1""";                  ValueType: string;  ValueName: ""; Components: client/pkmn
 
+Root: HKCR; Subkey: ".apladx";                                    ValueData: "{#MyAppName}ladxpatch";        Flags: uninsdeletevalue; ValueType: string;  ValueName: ""; Components: client/ladx
```
Comment (Berserker66) [inline inno_setup.iss:224]: unfortunately this is not enough to get the baserom integrate dfor generation and patching. Unless that is not needed for this game?

File: worlds/ladx/docs/setup_en.md (hunk 12 above)
Comment (zig-for) [reply worlds/ladx/docs/setup_en.md]: Fixed

File: worlds/ladx/docs/en_Links Awakening DX.md [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+## I don't know what to do!
+
+That's not a question - but I'd suggest clicking the crow icon on your client, which will load an AP compatible autotracker for LADXR.
+
+## What is this randomizer based on?
+
+This randomizer is based on (forked from) the wonderful work daid did on LADXR - https://github.com/daid/LADXR
+
+The autotracker code for communication with magpie tracker is directly copied from kbranch's repo - https://github.com/kbranch/Magpie/tree/master/autotracking
```
[the remark this answers is not in the document]
Comment (zig-for) [reply worlds/ladx/docs/en_Links Awakening DX.md:43]: I'll get kbranch to supply a license. He gave permission, I promise. :) Good catch.

File: worlds/ladx/docs/setup_en.md [hunk 15]
```diff
@@ -0,0 +1,93 @@
+# Links Awakening DX Multiworld Setup Guide
+
+## Required Software
+
+- [Archipelago](https://github.com/ArchipelagoMW/Archipelago/releases). Make sure to check the box for `Links Awakening DX`
+- Hardware or software capable of loading and playing SNES ROM files
```
[the remark this answers is not in the document]
Comment (zig-for) [reply worlds/ladx/docs/setup_en.md:6]: dangit

File: worlds/ladx/docs/en_Links Awakening DX.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (zig-for) [reply worlds/ladx/docs/en_Links Awakening DX.md:43]: fwiw - LADXR is MIT licensed. https://github.com/daid/LADXR/blob/master/LICENSE

File: worlds/ladx/docs/setup_en.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (zig-for) [reply worlds/ladx/docs/setup_en.md]: fixed

File: inno_setup.iss (hunk 13 above)
Comment (zig-for) [reply inno_setup.iss:227]: I think this is fixed now

File: worlds/ladx/docs/en_Links Awakening DX.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (zig-for) [reply worlds/ladx/docs/en_Links Awakening DX.md:43]: Resolved.

[the remark this answers is not in the document]
Comment (kbranch) [reply worlds/ladx/docs/en_Links Awakening DX.md:43]: Zig definitely asked for permission before touching anything, for what it's worth. I thought I had added an MIT LICENSE file, but apparently I forgot it - I just added it.

Magpie also doesn't use any actual code from the old Emotracker pack, so anything he uses is safe. If there are still concerns, I'm sure MuffinJets would happily make it official in his repo as well.

File: LinksAwakeningClient.py [hunk 16]
```diff
@@ -0,0 +1,610 @@
...
+            pass
+        finally:
+            all_tasks.remove(task)
+    task = asyncio.create_task(_log_exception(awaitable))
+    all_tasks.add(task)
+
+
+class LinksAwakeningContext(CommonContext):
+    tags = {"AP"}
+    game = "Links Awakening DX"  # empty matches any game since 0.3.2
```
Comment (Berserker66) [inline LinksAwakeningClient.py:432]: these comments are copied from text clients and all 3 don't apply to this client

File: worlds/ladx/Options.py [hunk 17]
```diff
@@ -0,0 +1,384 @@
...
+
+class TradeQuest(DefaultOffToggle, LADXROption):
+    """
+    On - adds the trade items to the pool (the trade locations will always be local items)
+    Off - (default) doesn't add them
+    """
+    ladxr_name = "tradequest"
+
+
+#            Setting('forwardfactor', 'Main', 'F', 'Forward Factor', default=0.0,
```
Comment (Berserker66) [inline worlds/ladx/Options.py:45]: looks like leftover stuff that could be removed? I found a bunch of single line old code comments as well, but this one is a bit big

File: worlds/ladx/Options.py [hunk 18]
```diff
@@ -0,0 +1,384 @@
...
+    """
+
+    normal = 0
+    gift = 1
+    default = gift
+
+# TODO: translate to lttp parlance
+class EntranceShuffle(Choice, LADXROption):
+    """
+    [WARNING] Experimental, may break generation
```
Comment (Berserker66) [inline worlds/ladx/Options.py:74]: What does "break generation" mean in AP context?

File: worlds/ladx/__init__.py [hunk 19]
```diff
@@ -0,0 +1,421 @@
...
+    #    "weapons": {"sword", "lance"}
+    #}
+
+    prefill_dungeon_items = None
+
+    player_options = None
+
+    def create_item(self, item: str) -> LinksAwakeningItem:
+        assert(False)
+
```
Comment (Berserker66) [inline worlds/ladx/__init__.py:80]:
```suggestion
```
This is defined again further down, so this can go away.

File: worlds/ladx/__init__.py [hunk 20]
```diff
@@ -0,0 +1,421 @@
...
+        patch.write()
+        if not DEVELOPER_MODE:
+            os.unlink(rompath)
+
+    def generate_multi_key(self):
+        return bytes.fromhex(self.multiworld.seed_name) + self.player.to_bytes(2, 'big')
+
+    def modify_multidata(self, multidata: dict):
+        multi_key = binascii.hexlify(self.generate_multi_key()).decode()
+        multidata["connect_names"][multi_key] = multidata["connect_names"][self.multiworld.player_name[self.player]]
```
Comment (Berserker66) [inline worlds/ladx/__init__.py:421]: I see you encode a connect name, but also allow the client user entry of the player name?
I think that makes this the first game to do that, normally if a connect hash is encoded, that's the only way to connect with that game's client.

File: worlds/ladx/docs/en_Links Awakening DX.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (muffinjets) [reply worlds/ladx/docs/en_Links Awakening DX.md:43]: I can't speak for how EmoTracker as a core applies here, but if Magpie has any leftover code from my work, I give full permission and will apply MIT if needed

File: worlds/ladx/Options.py [hunk 21]
```diff
@@ -0,0 +1,384 @@
...
+class Logic(Choice, LADXROption):
+    """Affects where items are allowed to be placed.
+    [Casual] Same as normal, except that a few more complex options are removed, like removing bushes with powder and killing enemies with powder or bombs.
+    [Normal] Playable without using any tricks or glitches. Can require knowledge from a vanilla playthrough, such as how to open Color Dungeon.
+    [Hard] More advanced techniques may be required, but glitches are not. Examples include tricky jumps, killing enemies with only pots.
+    [Glitched] Advanced glitches and techniques may be required, but extremely difficult or tedious tricks are not required. Examples include Bomb Triggers, Super Jumps and Jesus Jumps.
+    [Hell] Obscure knowledge and hard techniques may be required. Examples include featherless jumping with boots and/or hookshot, sequential pit buffers and unclipped superjumps. Things in here can be extremely hard to do or very time consuming."""
+    display_name = "Logic"
+    ladxr_name = "logic"
+    # option_casual = 0
```
Comment (Alchav) [inline worlds/ladx/Options.py:29]: Option is described in description but is disabled

File: worlds/ladx/Options.py (hunk 18 above)
Comment (Berserker66) [reply worlds/ladx/Options.py:56]: I believe SM can do that too, in which case it forces accessibility minimal

**Review by Berserker66 — APPROVED** [review]
